src/Microsoft.Health.Fhir.TemplateManagement/OCIFileManager.cs

```diff
@@ -5,90 +5,111 @@
 
 using System.Collections.Generic;
 using System.IO;
-using System.Linq;
 using System.Threading.Tasks;
-using EnsureThat;
 using Microsoft.Health.Fhir.TemplateManagement.Client;
+using Microsoft.Health.Fhir.TemplateManagement.Exceptions;
 using Microsoft.Health.Fhir.TemplateManagement.Models;
 using Microsoft.Health.Fhir.TemplateManagement.Overlay;
 
 namespace Microsoft.Health.Fhir.TemplateManagement
 {
-    public class OCIFileManager
+    public class OciFileManager
     {
-        private readonly OrasClient _orasClient;
-        private readonly OverlayFileSystem _overlayFS;
-        private readonly OverlayOperator _overlayOperator;
+        private readonly IOciClient _client;
+        private readonly IOverlayFileSystem _overlayFS;
+        private readonly IOverlayOperator _overlayOperator;
 
-        public OCIFileManager(string imageReference, string workingFolder)
+        public OciFileManager(string registry, string workingFolder)
         {
-            ImageInfo.ValidateImageReference(imageReference);
-            _orasClient = new OrasClient(imageReference);
             _overlayFS = new OverlayFileSystem(workingFolder);
+            _client = new OrasClient(registry, Path.Combine(workingFolder, Constants.HiddenLayersFolder));
             _overlayOperator = new OverlayOperator();
         }
 
-        public async Task PullOCIImageAsync()
+        /// <summary>
+        /// Pull image layers into working folder.
+        /// Layers will be written into hidden image folder.
+        /// </summary>
+        /// <param name="name">Image name.</param>
+        /// <param name="reference">Image reference (digest or tag).</param>
+        /// <param name="forceOverride">Whether to override the existed files.</param>
+        /// <returns>OCI artifact image.</returns>
+        public async Task<ArtifactImage> PullOciImageAsync(string name, string reference, bool forceOverride = false)
         {
-            _overlayFS.ClearImageLayerFolder();
-            await _orasClient.PullImageAsync(_overlayFS.WorkingImageLayerFolder);
+            if (!_overlayFS.IsCleanWorkingFolder() && !forceOverride)
+            {
+                throw new TemplateManagementException($"The folder is not empty. If force to override, please add -f in parameters");
```

Comment:
We may give some more detailed message. We will only overwrite files with the same name and will not delete other files and folders